bundles/org.openhab.binding.bluetooth.bluegiga/src/main/java/org/openhab/binding/bluetooth/bluegiga/internal/BlueGigaSerialHandler.java

```diff
@@ -294,23 +315,33 @@ private void removeTransactionListener(BluetoothListener<?> listener) {
      * @param bleCommand {@link BlueGigaCommand}
      * @return response {@link Future} {@link BlueGigaResponse}
      */
-    public <T extends BlueGigaResponse> Future<T> sendBleRequestAsync(final BlueGigaCommand bleCommand,
+    private <T extends BlueGigaResponse> Future<T> sendBleRequestAsync(final BlueGigaCommand bleCommand,
             final Class<T> expected) {
         checkIfAlive();
         class TransactionWaiter implements Callable<T>, BluetoothListener<T> {
-            private boolean complete;
+            private volatile boolean complete;
             private BlueGigaResponse response;
+            BlueGigaUniqueCommand query = new BlueGigaUniqueCommand(bleCommand, transactionId.getAndIncrement());
 
             @SuppressWarnings("unchecked")
             @Override
-            public T call() {
+            public T call() throws TimeoutException {
                 // Register a listener
                 addTransactionListener(this);
 
                 // Send the transaction
-                queueFrame(bleCommand);
+                queueFrame(query);
+
+                // Start transaction timeout timer
+                Future<?> timeout = executor.schedule(() -> {
+                    complete = true;
+                    logger.debug("Timeout, no response received for transaction {}", query.getTransactionId());
+                    synchronized (this) {
+                        notify();
+                    }
+                }, TRANSACTION_TIMEOUT_PERIOD_MS * 2, TimeUnit.MILLISECONDS);
```

Comment:
If `executor` is single threaded, how is this timeout ever supposed to run if the single thread is going to be busy with this task? 🤔